### ISSUE TITLE
Question: How best to support netCDF?

### ISSUE THREAD

**Issue by cpaulik** [EOxServer/eoxserver#310]
We at TU-Wien would be very interested in having support for netCDFs formatted according to the [CF conventions](http://cfconventions.org/).

I would be willing to work on this but need your help in figuring out the best/most realistic point of attack for this problem. I see several possibilities, but there might be others: 
- GDAL (a little unrealistic at the moment for me since I would need to learn how to contribute to GDAL)
- Mapserver (same thing as above)
- EOXServer (more realistic since it is in Python which I already know well. But depends on how the integration with Mapserver takes place)

At the Python level there are already at least two packages, namely [xray](http://xray.readthedocs.org/en/stable/) and [iris](https://github.com/SciTools/iris), that are able to work with DataCube like netCDF files with several variables. I would prefer xray here since it has a clearer purpose than iris and less dependencies.

I guess my basic question is if it would be possible to use e.g. xray as a backend for netCDF files in EOXserver? 

Or do you have any other plans in this direction?

**Comment by Schpidi** [comment]
Hi @cpaulik, great to hear from you :) In fact we've started to discuss how to add netCDF support to EOxServer and would be very interested to learn what your needs are. Since you're also based in Vienna what do you think about a joint coding session? cc @constantinius

**Comment by constantinius** [comment]
Hi @cpaulik!
It's nice to see that there is interest in extending EOxServer!

Currently, our data-structure is inclined towards 2D raster data, projected onto the Earths surface. Within EOxServers core, we did not really deal with objects of higher dimensionality, though we are very interested in extending the current capabilities. So your help is very welcome!

Coincidentally, we talked about supporting the handling of arbitrary dimensions and formats beyond GDAL.

To hook in where @Schpidi left off: what are your desired features? Do you want to visualize the netCDF data in the context of a map? Then we would have to figure out a way to hook this into the WMS rendering process. Are you interested into delivering the actual data values? WCS would be the best way, and netCDF a very good format to deliver n-D datasets. Do you want to hook in a server-side processing? WPS would be your friend.

Regarding xray or iriss: I have no experience with either of the two libraries, but xray did come up just recently in another context. I think you are free to choose whatever suits best.

I think a joint coding session would be a good way to discuss that matter!

**Comment by cpaulik** [comment]
Glad to hear that you are interested.

Our main interest at the moment is the extraction of images (WMS) and the extraction of time series of a grid point (lat, lon) so a 1D slice via WCS would do just fine. 

WPS or manipulation of the actual values would also be nice for e.g. interactive masking of a dataset or maybe even simple layer math, but my main focus at the moment is getting the first two working.

A joint coding session sounds very good to me. I'll setup a EOXServer dev environment on my laptop and will read the docs in more detail. Would the first or second week of November work for you?

**Comment by Schpidi** [comment]
Sorry for my delayed answer. What about Monday or Tuesday next week? Do you want to visit us in our office? We'll provide essentials like Coffee and WiFi :)

**Comment by cpaulik** [comment]
Next Tuesday works for me. Yes I can visit your office. Would 10 AM be OK for you?

**Comment by Schpidi** [comment]
Perfect cu on Tuesday

**Comment by cpaulik** [comment]
At the meeting we discovered a few issues that might be raised individually:

- [ ] Reading NetCDF data like the [ESA CCI dataset](http://www.esa-soilmoisture-cci.org/) does not work because GDAL does not handle the `scale_factor` correctly, maybe even not at all.
  - EOXServer RangeTypes provide no solution for this at the moment but might be an option.
- [ ] Reading netCDF data of TU Wien internal Sentinel 1 data  `SAR.nc` on EQUI-7 projection did not work because of missing ESPG code of the projection.
  - A solution would be to support custom projections in EOXServer e.g. PROJ4 Strings
- [ ] Time series extraction could be tackled through WPS from what I understand but only after the datasets are properly supported.

**Comment by Schpidi** [comment]
Some further comments:
* We succeeded in registering ESA CCI data after changing the datatype to float32 using `ncap` from the NCO package. Seems there is an issued with GDAL which we might want to investigate.
* The issues with non EPSG projections like in the TU Wien internal Sentinel 1 data was fixed in 2b66e63274c33d6a30747716321898ee4bf71b53 and PROJ4 strings should work now.

According to my understanding looking at time series extraction is now the next step isn't it?
